PyPDF2/generic/_data_structures.py

```diff
@@ -387,32 +499,61 @@ def addChild(self, child: Any, pdf: Any) -> None:  # pragma: no cover
+        self.insert_child(child, None, pdf)
+
+    def insert_child(self, child: Any, before: Any, pdf: Any) -> None:  # PdfWriter
+        def inc_parent_counter(
+            parent: Union[None, IndirectObject, TreeObject], n: int
+        ) -> None:
+            if parent is None:
+                return
+            parent = cast("TreeObject", parent.get_object())
+            if "/Count" in parent:
+                parent[NameObject("/Count")] = NumberObject(
+                    cast(int, parent[NameObject("/Count")]) + n
+                )
+                inc_parent_counter(parent.get("/Parent", None), n)
+
         child_obj = child.get_object()
-        child = pdf.get_reference(child_obj)
-        assert isinstance(child, IndirectObject)
+        child = child.indirect_ref  # get_reference(child_obj)
+        # assert isinstance(child, IndirectObject)
 
         prev: Optional[DictionaryObject]
-        if "/First" not in self:
+        if "/First" not in self:  # no child yet
             self[NameObject("/First")] = child
             self[NameObject("/Count")] = NumberObject(0)
-            prev = None
+            self[NameObject("/Last")] = child
+            child_obj[NameObject("/Parent")] = self.indirect_ref
+            inc_parent_counter(self, child_obj.get("/Count", 1))
+            if "/Next" in child_obj:
+                del child_obj["/Next"]
+            if "/Prev" in child_obj:
+                del child_obj["/Prev"]
+            return
         else:
-            prev = cast(
-                DictionaryObject, self["/Last"]
-            )  # TABLE 8.3 Entries in the outline dictionary
-
-        self[NameObject("/Last")] = child
-        self[NameObject("/Count")] = NumberObject(self[NameObject("/Count")] + 1)  # type: ignore
-
-        if prev:
-            prev_ref = pdf.get_reference(prev)
-            assert isinstance(prev_ref, IndirectObject)
-            child_obj[NameObject("/Prev")] = prev_ref
-            prev[NameObject("/Next")] = child
-
-        parent_ref = pdf.get_reference(self)
-        assert isinstance(parent_ref, IndirectObject)
-        child_obj[NameObject("/Parent")] = parent_ref
+            prev = cast("DictionaryObject", self["/Last"])
+
+        while prev.indirect_ref != before:
+            if "/Next" in prev:
+                prev = cast("TreeObject", prev["/Next"])
+            else:  # append at the end
+                prev[NameObject("/Next")] = cast("TreeObject", child)
+                child_obj[NameObject("/Prev")] = prev.indirect_ref
+                child_obj[NameObject("/Parent")] = self.indirect_ref
+                if "/Next" in child_obj:
+                    del child_obj["/Next"]
+                self[NameObject("/Last")] = child
+                inc_parent_counter(self, child_obj.get("/Count", 1))
+                return
+        try:  # insert as first or in the middle
+            prev["/Prev"][NameObject("/Next")] = child  # type: ignore
```

Comment:
```suggestion
            assert isinstance(prev["/Prev"], DictionaryObject)
            prev["/Prev"][NameObject("/Next")] = child
```

Reply:
is is OK for performance ?

Reply:
I don't think it's a big deal. On the one hand, people could remove that by using `python -O` (which I haven't ever seen). On the other hand, the `isinstance` call is cheap:

```
def test_exception():
    try:
        a = 1/0
        return a
    except Exception:
        return 2


if __name__ == '__main__':
    import timeit
    print(f'try-except        : {timeit.timeit("test_exception()", setup="from __main__ import test_exception", number=10**6)*1000:3.0f}ns')
    print(f'isinstance false  : {timeit.timeit("isinstance(foo, int)", setup="""foo="a" """, number=10**6)*1000:3.0f}ns')
    print(f'isinstance true   : {timeit.timeit("isinstance(foo, str)", setup="""foo="a" """, number=10**6)*1000:3.0f}ns')
    print(f'addition int-int  : {timeit.timeit("123456+9876554", number=10**6)*1000:3.0f}ns')
    print(f'addition int-float: {timeit.timeit("123456+9876554.3", number=10**6)*1000:3.0f}ns')
```

gives:

```
try-except        : 287ns
isinstance false  :  98ns
isinstance true   :  73ns
addition int-int  :   9ns
addition int-float:   9ns
```

Our big performance problem is how we deal with floats. The conversion of floats to FloatObjects is dominating pretty much everything.